### ISSUE TITLE
Brenda Tissue Ontology terms aren't suggesting correctly

### ISSUE THREAD

**Issue by jseager7** [pombase/canto#2183]
Since a recent ontology reload, Canto suggests different terms when searching for the BTO term 'leaf':

![image](https://user-images.githubusercontent.com/37659591/72810041-bd5b0600-3c54-11ea-9663-d5eb8e05905f.png)

The closest (and intended) match for this query is 'leaf' [(BTO:0000713)](https://www.ebi.ac.uk/ols/ontologies/bto/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FBTO_0000713), and that was previously included in term suggestions.

I'm guessing this is because 'leaf' has a `part_of` relation to other terms, rather than an `is_a` relation:

> **id:** BTO:0000713
> **name:** leaf
> **namespace:** BrendaTissueOBO
> **def:** "A lateral outgrowth from a plant stem that is typically a flattened expanded variably shaped greenish organ, constitutes a unit of the foliage, and functions primarily in food manufacture by photosynthesis." [From_Merriam-Webster's_Online_Dictionary_at_www.Merriam-Webster.com:http\://www.m-w.com/cgi-bin/dictionary?book=Dictionary&va=leaf]
> **relationship:** part_of BTO:0001243 ! shoot
> **relationship:** part_of BTO:0001495 ! plumule

Our annotation extension configuration file only specifies `part_of` as a relation, but according to the wiki, this subset relation only applies to the _domain_ ontology (PHIPO), not the _range_ ontology (BTO), so this might not be anything to do with the problem.

domain ID | subset relation | extension relation | range ID | Canto display text | Help text | cardinality | role
-- | -- | -- | -- | -- | -- | -- | --
PHIPO:0000001 | is_a | infects_tissue | BTO:0000000 | host tissue infected | (help text) | * | user

@kimrutherford Could this simply be a bug with the term lookup system?

**Comment by ValWood** [comment]
I would guess that if we use is_a and part of we should  be able to access everything.
Although I thought that everything should have an is_a relationship?

**Comment by jseager7** [comment]
> I would guess that if we use is_a and part of we should be able to access everything. Although I thought that everything should have an is_a relationship?

Everything in PHIPO is related with `is_a` (as far as I know), but the problem seems to be that the term lookup for the _range_ ontology (BTO) isn't able to cope with `part_of` relations. I've tested this by looking up terms from BTO that have a normal subclass relation, compared to 'subclass of part of', and every example of the latter doesn't return results in Canto, whereas examples of the former do.

Basically, the term suggestion only seems to be taking the `is_a` hierarchy into account.

**Comment by jseager7** [comment]
I could test changing the subset relation to `part_of` or a conjunction of `is_a|part_of` (if that's even possible), but the documentation specifically states that the subset relation only matches relations in the domain ontology (PHIPO):

> subset relation - The relation used to find domain term descendants (usually "is_a")

**Comment by kimrutherford** [comment]
>  could test changing the subset relation to part_of, or a conjunction of is_a|part_of (if that's even possible),

I believe "is_a|part_of" is possible.  Could you try it?  If it works for you I'll update the documentation to mention it.  :-)

**Comment by jseager7** [comment]
> I believe "is_a|part_of" is possible. Could you try it?

I'm not sure exactly how I can prove that this works, short of creating a test ontology just for this purpose, because PHIPO has no `part_of` relations. I could try and run an experiment locally with a minimal test ontology if you want me to test this regardless.

Just to clarify, the problem I'm trying to solve is that terms from the _range_ ontology (BTO) that are related by `part_of` are not showing up in the term suggestion. There's no option in the config to specify a subset relation for the range ontology; you can only specify subset relations for the domain ontology.

**Comment by ValWood** [comment]
What is "range ontology" and "domain ontology" in this context?

**Comment by jseager7** [comment]
> What is "range ontology" and "domain ontology" in this context?

I'm using the terminology from the annotation extension configuration, specifically the tab-separated value (TSV) file. In the example below, you can see the columns 'Domain ID' and 'Range ID'. They're analogous to the domain and range of a mathematical function:

domain ID | subset relation | extension relation | range ID | Canto display text | Help text | cardinality | role
-- | -- | -- | -- | -- | -- | -- | --
PHIPO:0000001 | is_a | infects_tissue | BTO:0000000 | host tissue infected | (help text) | * | user

The **domain ontology** specifies the terms to which you can apply an annotation extension. In this case, it means that only the children of PHIPO:0000001 can be used as the 'starting point' for the infects_tissue relation. The **range ontology** is the set of terms (or values) that can be selected as the 'end point' for the relation: those are the options you can pick from when creating the annotation extension.

So, infects_tissue is a relation that has PHIPO terms as its domain and BTO terms as its range. The domain ontology is PHIPO and the range ontology is BTO.

**Comment by ValWood** [comment]
OK, I understand  but usually   domain and range refer to the same ontology.

But this just means that  (presumably) andy BTO tissue ID can be used  with any PHIPO term (I.e both are specifying root nodes). So I think the fix is just to include part_of in the subset relation?

**Comment by jseager7** [comment]
> So I think the fix is just to include part_of in the subset relation?

Unfortunately the annotation extension documentation states that the subset relation only applies to the domain ontology ([link](https://github.com/pombase/canto/wiki/AnnotationExtensionConfig)). There doesn't seem to be anything I can do to specify a subset relation for the range ontology, so I can't get Canto to include part_of terms from BTO (and most of the terms are related with part_of). We'd need an extra column for 'range subset relation' or something.

**Comment by ValWood** [comment]
OK got it....I gues this would not be a problem if BTO was is_a complete... that's possibly why we never had the issue before?
@mah11  is that right?

**Comment by mah11** [comment]
>  but usually domain and range refer to the same ontology.

this isn't true

> I guess this would not be a problem if BTO was is_a complete

probably, yes

**Comment by kimrutherford** [comment]
> Just to clarify, the problem I'm trying to solve is that terms from the range ontology (BTO) that
> are related by part_of are not showing up in the term suggestion. There's no option in the
> config to specify a subset relation for the range ontology; you can only specify subset relations
> for the domain ontology.

Ah, right.  I've got it now.  :-)

Yep, there's no option at the moment.  I'd prefer no to add one if we can avoid it at it makes the config (and code) even more complicated.

> I'm guessing this is because 'leaf' has a part_of relation to other terms, rather than an is_a relation:

Sorry I should have thought more before answering.  I don't think that the relations between terms should make a difference to the autocomplete as it only looks at names and synonyms.  Could you send me a link to the BTO OBO file you're loading so I can try this locally?

**Comment by jseager7** [comment]
> Could you send me a link to the BTO OBO file you're loading so I can try this locally?

We're loading this version of BTO:

https://raw.githubusercontent.com/BRENDA-Enzymes/BTO/master/bto.obo

There's also other release formats on the repository, but most of them seem to be equivalent to bto.obo. The bto-simple.obo file is practically empty.

**Comment by kimrutherford** [comment]
> I'm guessing this is because 'leaf' has a part_of relation to other terms, rather than an is_a relation:

Thanks for the link to BTO.  I think you're right - the problem is that leaf has no is_a parents.  Because of that the loading code puts "leaf" in the internal canto_root_subset.  Terms in that subset are ignored during extension autocomplete.

The default configuration is:

```
 ontology_namespace_config:
   subsets_to_ignore:
     primary_autocomplete:
       - "is_a(Grouping_terms)"
     primary_select:
       - "is_a(Grouping_terms)"
       - "is_a(qc_do_not_annotate)"
     extension:
       - "is_a(canto_root_subset)"
       - "is_a(qc_do_not_annotate)"
   do_not_annotate_subsets:
     - "is_a(canto_root_subset)"
     - "is_a(qc_do_not_annotate)"
```

We might be able to change the loading code so that terms with a part_of parent aren't added to canto_root_subset.  That may change the root subset for other ontologies but should be fine for FYPO and GO at least.

**Comment by jseager7** [comment]
> That may change the root subset for other ontologies but should be fine for FYPO and GO at least.

It won't affect PHIPO either, since we don't use part_of.

Could the config add part_of to the list of subsets to ignore? Then there would be no harm in including part_of relations by default. For example:

```
primary_autocomplete:
  - "is_a(Grouping_terms)"
  - "part_of(Grouping_terms)"

**Comment by kimrutherford** [comment]
> For example:

I don't think that would help in your case.  It's the
```
subsets_to_ignore -> extension -> "is_a(canto_root_subset)"
```
setting that's causing this.

Let's have a chat about this on Skype.

**Comment by kimrutherford** [comment]
James, we should talk about this on the next call.  I'm not sure there is an easy fix.  It's frustrating that BTO doesn't have a top level term like "tissue".  Having "leaf" at the top level seems a bit strange.

**Comment by kimrutherford** [comment]
> It's frustrating that BTO doesn't have a top level term like "tissue". Having "leaf" at the top
> level seems a bit strange.

I take that back.  "leaf" isn't top level because although it has no is_a parents, it's part_of(shoot) and part_of(plumule).  It's only top-level if consider just is_a relations, which is what Canto does.

**Comment by jseager7** [comment]
Is it feasible for Canto to use a different relation as its 'dominant' relation, on a per-ontology basis? I don't know how common it is for ontologies to be structured with `part_of`, but I suspect BTO won't be the only one.

**Comment by mah11** [comment]
> Is it feasible for Canto to use a different relation as its 'dominant' relation, on a per-ontology basis?

I really don't recommend spending any time or effort trying to do this. The "dominance" of is_a in Canto reflects how the relation is used in OBO ontologies (corresponding to the SubClassOf relation in OWL). Diverging from that feels dangerous.

> I don't know how common it is for ontologies to be structured with part_of, but I suspect BTO won't be the only one.

As far as I know it's not rare for ontologies to use the part_of relation, but beyond that I don't think I fully understand what you mean by "structured with part_of". BTO is a bit unusual among the OBO Foundry ontologies in lacking is_a paths to the root for some terms. In an ideal world the BTO maintainers would fill in the missing is_a links; in real life I don't know if they have the resources to make that a priority.

**Comment by jseager7** [comment]
> BTO is a bit unusual among the OBO Foundry ontologies in lacking is_a paths to the root for some terms.

That's what I meant by 'structured', in that I presumed there was some way to reach every term by a `part_of` relation, but not by `is_a`, meaning `part_of` is like the backbone of the ontology. I couldn't think of any more precise wording at the time.

**Comment by ValWood** [comment]
Can't the config file be used to allow both part_of and is_a terms?
Presumably it is cardinality 1 at the moment  but could that not be extended to is_a,part_of ?

**Comment by mah11** [comment]
> Can't the config file be used to allow both part_of and is_a terms?

This was addressed [earlier in this ticket](https://github.com/pombase/canto/issues/2183#issuecomment-579220434).

**Comment by mah11** [comment]
> That's what I meant by 'structured', in that I presumed there was some way to reach every term by a part_of relation, but not by is_a, meaning part_of is like the backbone of the ontology.

Ah, ok, thanks for the clarification. In that case, it is indeed rare. Even BTO uses is_a to organize its most general levels; it just also has lots of more specific terms that are only connected to the basic stucture by part_of (it isn't what GO folks would call "is_a-complete"). IOW it's not that BTO uses part_of instead of is_a fundamentally.

**Comment by ValWood** [comment]
I  re-read and I still don't understand why for BTO  extensions the config file cannot specify is_a,part_of

**Comment by jseager7** [comment]
I re-read and I still don't understand why for BTO extensions the config file cannot specify is_a,part_of

> We'd need a second 'subset relation' column in the extension config file that applies to the range ontology (BTO). The current 'subset relation' column only tells Canto how to select terms from PHIPO. The range ontology (BTO) always defaults to `is_a`.

**Comment by jseager7** [comment]
If we had a second 'subset relation' column for the range ontology, we could do this:

domain ID | domain relation | extension relation | range ID | range relation | Canto display text | Help text | cardinality | role
-- | -- | -- | -- | -- | -- | -- | -- | --
PHIPO:0000001 | is_a | infects_tissue | BTO:0000000 | is_a\|part_of | host tissue infected | (help text) | * | user

**Comment by ValWood** [comment]
AH got it! 
Can we cheat? 
Substitute part_of for is_a on import. We don't  displaying  these relations or doing anything which would confuse users.

I guess it depends how big an effect it would be changing range relation to  to is_a|part_of more generally, and I don't have a good feel for that at all.

**Comment by jseager7** [comment]
> AH got it!
Can we cheat?
Substitute part_of for is_a on import. We don't displaying these relations or doing anything which would confuse users.

I'll have to defer to @kimrutherford but I suspect adding a column would be simpler than any cheat to work around the problem, because the configuration table above is processed at the same time that the ontologies are loaded.

> I guess it depends how big an effect it would be changing range relation to to is_a|part_of more generally, and I don't have a good feel for that at all.

If it would fix the term suggestion problem that this issue was originally about (see https://github.com/pombase/canto/issues/2183#issue-552906311), then that would be a pretty big effect.

However, I can't really anticipate what the child term suggestion process would be like if we're also showing **part_of** relations there. We'd probably want some way for the interface to indicate which child terms are related by **part_of** and which are related by **is_a**, otherwise it could get confusing.

**Comment by kimrutherford** [comment]
> I'll have to defer to @kimrutherford but I suspect adding a column would be simpler than
> any cheat to work around the problem,

I think you're probably right that a new column for "range subset relation".  At the moment the code uses the domain subset relation as the range subset relation.  That's not a great work by me.  Luckily the domain subset relation is always "is_a" so it all works fine, except for BTO.

Adding a new column and implementing a range subset relation is a bit of a pain so I'm going to have a think before going ahead.  I'm hoping there's a simpler solution but I'm doubtful.

**Comment by jseager7** [comment]
There's now an issue open on the BTO repository, at https://github.com/BRENDA-Enzymes/BTO/issues/10, that proposes fixing the class hierarchy in BTO itself.

**Comment by jseager7** [comment]
I've tried the simple fix of forcing all of the relations to is_a, and it seems to work well, providing you view the child terms as more specific versions of the parents. Shown below is how 'leaf' fits into the hierarchy now. I'm planning to test this fix on the demo server soon.

![bto_coerce_is-a](https://user-images.githubusercontent.com/37659591/82906172-c0e0ff80-9f5c-11ea-800d-2834d6584ed5.png)

**Comment by kimrutherford** [comment]
Thanks James.  I think that's probably a good quick fix for now.

**Comment by jseager7** [comment]
The modified version of BTO is loaded on the demo server now, and it looks like the suggestions work as expected again, or maybe even better than they did originally:

![image](https://user-images.githubusercontent.com/37659591/83008136-9fd7e780-a00c-11ea-87a5-10ca9b4c8d6b.png)

The child term suggestions are working too:

![image](https://user-images.githubusercontent.com/37659591/83008370-f513f900-a00c-11ea-920b-be46cc470364.png)

The child suggestions are sometimes unexpected – for example 'bile' (BTO:0000121) as a more specific term for 'liver' (BTO:0000759) – but the only bad thing about this is that it clutters up the suggestion list with terms that might be rarely used. However, it might be better for users if we show everything that _could_ be related to the parent term, in which case this isn't a problem at all.

**Comment by jseager7** [comment]
Before l load this on the production server, I'd like to get the ontology loading problems in issue #2288 fixed first, providing it's not going to take a long time to fix the problem.

**Comment by jseager7** [comment]
This is fixed on both servers now, so I'm going to close this issue. If anyone has any problems with the way the ontology works now (@CuzickA in particular), feel free to re-open.
